Add Blackjack app example (#15202)
MystenLabs/sui: docs/content/guides/developer/app-examples/blackjack.mdx

---
title: Blackjack
---

> The following documentation goes through an example implementation of the popular casino game blackjack on Sui. This guide will walk through its components, providing a detailed look at the module's functions, structures, constants, and their significance in the overall gameplay mechanism.

Reviewer: ```suggestion
The following documentation goes through an example implementation of the popular casino game blackjack on Sui. This guide walks through its components, providing a detailed look at the module's functions, structures, constants, and their significance in the overall gameplay mechanism.
```

> Building an on-chain blackjack game shares a lot of similarities with a coin flip game, For that reason, this example will only cover the smart contracts (Move modules) and frontend logic.

Reviewer: ```suggestion
Building an on-chain blackjack game shares a lot of similarities with a coin flip game. For that reason, this example covers only the smart contracts (Move modules) and frontend logic.
```

:::info

For more details on building a backend and deploying to Sui, check out the [Coin Flip app example](./coin-flip.mdx).

You can also find the [full repository for this example here](https://github.com/MystenLabs/blackjack-sui).

:::

## Gameplay
In this single-player version of blackjack, the player competes against a dealer, which is automated by the system. The dealer is equipped with a public BLS key that plays a central role in the game's mechanics. Players generate randomness for the game by interacting with their mouse on the screen, after which they place their bet to start the game. Upon initiating the game, the dealer's backend, actively listening for the start transaction, processes it by signing and subsequently dealing two cards to the player and one to itself.

> The player has the options to 'Hit' or 'Stand.' Selecting 'Stand' triggers the dealer to draw cards until the total reaches 17 or higher. Once the dealer stops, typically at a sum of 17 or more, the smart contract steps in to compare the totals and declare the winner. On the other hand, choosing 'Hit' prompts the dealer to draw an additional card for the player.

Reviewer: ```suggestion
The player has the options to 'Hit' or 'Stand.' Selecting 'Stand' triggers the dealer to draw cards until the total reaches 17 or higher. After the dealer stops, typically at a sum of 17 or more, the smart contract steps in to compare the totals and declare the winner. On the other hand, choosing 'Hit' prompts the dealer to draw an additional card for the player.
```

## Smart contracts

> 

Reviewer: Technically shouldn't have headings together with no text betwixt.

### UI components and styling

- **Card Display:** Cards are rendered using a combination of JSX and CSS, displaying the suit symbols and values, with conditional styling (`getSuitColor`) based on the card suit.

- **Game Status Updates:** The UI provides immediate feedback on game events, such as a bust or game completion, using toasts (`toast`) and loading indicators (`Hearts` from `react-loader-spinner`).

### User interaction and feedback

- **Game Creation and Exploration:** Players can start new games using a dedicated button that triggers the `openNewGameModal` function. Links are also provided for players to view game details on the Sui explorer, enhancing transparency and engagement.

- **Mouse Movement Tracking:** The module includes functionality to track and utilize mouse movements, contributing to the game's randomness generation, which is crucial for fair play and unpredictability.

## Comparison: Blackjack and Coin Flip

### Similarities

- **Blockchain-Based Logic:** Both games are built on Sui, leveraging its capabilities for decentralized applications. The core game logic for each resides in Move modules, ensuring secure and verifiable gameplay.

> - **State Management:** In both games, state management is crucial. For Blackjack, this involves managing the player and dealer's hands and scores using React state hooks. In Satoshi Coin Flip, the state is managed through Move structs like `HouseData`, which track the house's balance and other game-related details.

Reviewer: ```suggestion
- **State Management:** In both games, state management is crucial. For blackjack, this involves managing the player and dealer's hands and scores using React state hooks. In Satoshi Coin Flip, the state is managed through Move structs like `HouseData`, which track the house's balance and other game-related details.
```
It's blackjack elsewhere on this page. Not sure which is better but consistency is key.

- **Randomness and Fair Play:** Both games emphasize randomness for fairness. Blackjack uses a Counter NFT and player mouse movements to generate randomness, while Satoshi Coin Flip only uses a Counter NFT as a unique input for the [Verifiable Random Function (VRF)](../cryptography/ecvrf.mdx) in each game.

> - **Smart Contract Interactions:** Each game involves smart contract interactions for game actions like placing bets, dealing cards (Blackjack), or making guesses (Coin Flip). These interactions are crucial for executing the game's logic on the blockchain.

Reviewer: blackjack?

### Differences

- **Game Mechanics and Complexity:** Blackjack is a more complex game with multiple actions (hit, stand, deal) and state updates, requiring a more dynamic frontend. In contrast, Satoshi Coin Flip has a simpler mechanic centered around a single bet and guess outcome.

> - **User Interface (UI) Complexity:** The Blackjack game involves a more intricate UI to display cards, manage game states, and player interactions. Satoshi Coin Flip, being simpler in gameplay, requires a less complex UI.

Reviewer: I think I prefer Blackjack now, for whatever that's worth.

- **Backend Processing:** In Blackjack, the dealer is automated (the machine), and the player's actions directly influence game outcomes. In the Coin Flip game, the house (smart contract) plays a more passive role, primarily in initializing and finalizing the game based on the player's guess.

> - **Module Structure and Focus:** Blackjack game focuses more on frontend interactions and real-time updates. The Satoshi Coin Flip game, however, delves deeper into backend logic, with structures like `HouseCap` and `house_data` for initializing and managing game data securely on the blockchain.

Reviewer: ```suggestion
- **Module Structure and Focus:** The Blackjack game focuses more on frontend interactions and real-time updates. The Satoshi Coin Flip game, however, delves deeper into backend logic, with structures like `HouseCap` and `house_data` for initializing and managing game data securely on the blockchain.
```

- **Multi-Version Implementation:** The Satoshi Coin Flip game mentions two versions – one susceptible to MEV attacks and another that is resistant, indicating a focus on security and user experience variations. Such variations aren't implemented in Blackjack.

:::info

> The complete app example can be found in [the repo here](https://github.com/MystenLabs/blackjack-sui).

Reviewer: ```suggestion
The complete app example can be found in [the balackjack-sui repo](https://github.com/MystenLabs/blackjack-sui).
```
